clang/test/CodeGen/builtins-nvptx.c

```diff
@@ -294,6 +294,81 @@ __device__ void nvvm_atom(float *fp, float f, double *dfp, double df, int *ip,
   // CHECK: call i32 @llvm.nvvm.atomic.load.dec.32.p0i32
   __nvvm_atom_dec_gen_ui(uip, ui);
 
+  // CHECK: call i32 @llvm.nvvm.ld.gen.i.volatile.i32.p0i32
+  __nvvm_volatile_ld_gen_i(ip);
+  // CHECK: call i32 @llvm.nvvm.ld.global.i.volatile.i32.p1i32
+  __nvvm_volatile_ld_global_i((__attribute__((address_space(1))) int *)ip);
```

Comment:
Interesting. The builtins are declared with the pointer in the generic AS, yet here we're passing a pointer to the global AS and it does get us the right intrinsic overload. This is convenient. 

The downside of this is that unless user explicitly casts the pointer to a specific AS, we'll likely have unnecessary ASC to the generic AS, as I've mentioned above. It's not a bug as accesses via generic AS are OK, but rather an optimization opportunity.

Reply:
I am not sure I follow your thinking. This is lowered into a PTX instruction that uses global (not generic) addressing. Are you saying that compiler will cast this pointer to generic AS and than back to global AS? Also, what does ASC stand for?

Reply:
> I am not sure I follow your thinking. 

My fault. I should not do code reviews at 1am. Please ignore the second paragraph.

> This is lowered into a PTX instruction that uses global (not generic) addressing. 

This is the "interesting" part. Because the argument has been explicitly marked as being in global AS, things do work exactly as intended. In a way you've created an equivalent of a function overload for the builtin for pointers in different AS and that may allow clang to accept more than it should.

> Are you saying that compiler will cast this pointer to generic AS and than back to global AS? 

Normally, yes. See https://godbolt.org/z/aG1bd45ez. 

For the calls of `load` the pointer gets cast to global AS and then back to generic, before passing it to a function. You can see it in the IR, as well. Here you're probably saved by the fact that you handle call generation yourself, but you do sort of break the rules in the process. I.e. builtins with generic AS pointer arguments should not accept pointers in the specific AS. The test above demonstrates that these builtins do.

I'm curious what's going to happen if you actually call `__nvvm_volatile_ld_global_i(ip);`.
Technically, it does match the builtin signature, so clang will accept this. The way builtin code gets called, it will likely end up calling `@llvm.nvvm.ld.global.i.volatile.i32.p0i32`  that would probably translate into `ld.volatile` instruction with a generic pointer as an argument. I do not think it's a desired outcome.

I think builtins should have signatures with the pointers in the matching AS. This way clang will catch attempts to pass a wrong-AS pointer to them.

> Also, what does ASC stand for?
`ASC` = `addrspacecast` LLVM IR instruction.

Reply:
> For the calls of load the pointer gets cast to global AS and then back to generic, before passing it to a function.

I don't see any such casts to global and back to generic in generated PTX. That is either in the godbolt example you linked or locally when I tried something similar with the new builtins.

> I'm curious what's going to happen if you actually call __nvvm_volatile_ld_global_i(ip);.
Technically, it does match the builtin signature, so clang will accept this. The way builtin code gets called, it will likely end up calling @llvm.nvvm.ld.global.i.volatile.i32.p0i32 that would probably translate into ld.volatile instruction with a generic pointer as an argument. I do not think it's a desired outcome.

That is exactly what happens.

> I think builtins should have signatures with the pointers in the matching AS. This way clang will catch attempts to pass a wrong-AS pointer to them.

I can not easily specify which AS the argument must be in when declaring builtins. While setting the address space numbers in builtin declarations works for CUDA, it does not work for OpenCL, which is our use case for this. I could try to throw an error in code generation if a builtin is called with AS that is not the same one it accepts.

Reply:
I would suggest that we merge this as it is to unblock other atomic-related PRs. 

When we figure out how to enable typecasting between target and OpenCL language address spaces we come back to this and change the builtins in the way that was suggested in this thread. Regardless of what we decide here this will need to be done for atomic read-modify-write builtins that have already been merged.

Reply:
I'm okay with that plan if @Artem-B has no objections.